src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -449,23 +452,44 @@ protected function load($file)
      */
     protected function getEnv($name)
     {
+        if (isset($this->resolving[$envName = "env($name)"])) {
+            throw new ParameterCircularReferenceException(array_keys($this->resolving));
+        }
         if (isset($this->envCache[$name]) || array_key_exists($name, $this->envCache)) {
             return $this->envCache[$name];
         }
-        if (0 !== strpos($name, 'HTTP_') && isset($_SERVER[$name])) {
-            return $this->envCache[$name] = $_SERVER[$name];
+        if (!$this->has($id = 'container.env_providers_locator')) {
+            $this->set($id, new ServiceLocator(array()));
+        }
+        if (!$this->getEnv) {
+            $this->getEnv = new \ReflectionMethod($this, __FUNCTION__);
+            $this->getEnv->setAccessible(true);
+            $this->getEnv = $this->getEnv->getClosure($this);
         }
-        if (isset($_ENV[$name])) {
-            return $this->envCache[$name] = $_ENV[$name];
+        $providers = $this->get($id);
+
+        if (false !== $i = strpos($name, ':')) {
+            $prefix = substr($name, 0, $i);
+            $localName = substr($name, 1 + $i);
+        } else {
+            $prefix = 'string';
+            $localName = $name;
         }
-        if (false !== $env = getenv($name)) {
-            return $this->envCache[$name] = $env;
+        $provider = $providers->has($prefix) ? $providers->get($prefix) : new EnvProvider($this);
+
+        $this->resolving[$envName] = true;
+        try {
+            if (null !== $env = $provider->getEnv($prefix, $localName, $this->getEnv)) {
+                return $this->envCache[$name] = $env;
+            }
+        } finally {
+            unset($this->resolving[$envName]);
         }
-        if (!$this->hasParameter("env($name)")) {
+        if (!$this->hasParameter($envName)) {
```

Comment:
> > forbid setting prefixed envs as parameters

> that is hard-coding the way namespaces are used in the current env provider. But other env providers may use trailing : in the way they want, so that default prefixed env might be required. I wouldn't close this possibility.

And what about a fallback lookup here? `"env(NAME_ONLY)"` that is. Im still not sure in what way setting a namespaced default makes sense, but indeed no real reason to close the possibility.

Reply:
> preserve nullable behavior

good idea, done